### DISCUSSION TITLE
DatabaseId use in github-release datasource cache breaks on GHE 3.1.x

### DISCUSSION THREAD

**Discussion by olira** [renovatebot/renovate#17117] [Suggest an Idea]
Hi everybody,

while I was testing the most recent release of renovate against our 3.1.x GHE instance, I noticed that we were no longer able to update dependencies against the github-release datasource.

Digging deeper, the issue was introduced in conjunction with the [caching layer](https://github.com/renovatebot/renovate/pull/15653) for GitHub releases which uses an GraphQL field [`databaseId` as `id`](https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/github-releases/cache/index.ts#L19). 

Before opening a ticket / PR I was wondering whether there is a specific reason why `databaseId` was used for `id` rather than the `id` field.

Unfortunately I do not know with which version of GHE `databaseId` will be introduced and I also don't know the companies migration plans.

Any guidance appreciated :)

**Comment by viceice** [comment]
//cc @zharinov Do you know something?

**Reply by zharinov** [reply]
The GraphQL `databaseId` is an equivalent for the REST `id`, so I renamed it for compatibility with our current codebase. GraphQL `id` field is the [Base64-encoded string](https://docs.github.com/en/graphql/reference/scalars#id) and I believe we didn't use it in Renovate before or after the caching layer introduction.

**Reply by olira** [reply]
That makes sense. I'll grab an excerpt of the GraphQL results tomorrow to see whether the value would still be usable.

**Reply by olira** [reply]
Looking at the GraphQL the base64 id would be something along the line of `MDc6UmVsZWFzZTEwNjEyMzU=` which is decoded to `07:Release1061235`. Not sure what the 07 is (probably some serial), but after the colon it seems to be the entity type as well as the databaseId. In terms of uniqueness,  the Base64 encoded value should be as good as the `databaseId`. If there is no cross referencing with the REST API Release Id anywhere, could we consider using `id` rather than `databaseId`?

**Reply by zharinov** [reply]
I don't think relying on internal id representation is the good idea. `07` is most likely a revision which means that other six representations can be found somewhere else. Also, it is almost guaranteed to change in the future, meaning high risk of unpredictable breakage.

**Reply by zharinov** [reply]
But if it doesn't work for your GHE version anyway, maybe it can be reasonable to conditionally use this workaround only for this specific versions.

**Comment by rarkins** [comment]
Does it work in GHE 3.2?

Is 3.1 still supported by GitHub?

**Reply by rarkins** [reply]
It is unsupported: https://docs.github.com/en/enterprise-server@3.1/admin/release-notes

**Reply by olira** [reply]
Yes, that is correct. At least not for mainstream. "Special" customers still get critical updates. 

Since the field has been introduced into GH public on [June 7th last year](https://docs.github.com/en/graphql/overview/changelog#schema-changes-for-2021-06-07) and 3.2 has been released 2021-09-09 chances are that it supports the field. 

Not sure when central IT will migrate though (I don't expect it to happen in the near future).

**Reply by rarkins** [reply]
We wouldn't normally do this type of change intentionally and don't aggressively end support for platforms, it was just an oversight. But in this  case I don't think we'll revert it for an officially EOS GHE release.

**Reply by olira** [reply]
I would never have assume that and I couldn't agree more. This change should definitely not be reverted. If we could solve the issue by something as simple as changing `databaseId` to `id` while still keeping integrity of the cache and not break any other functionality, would you think this is something doable?

**Reply by rarkins** [reply]
Are you proposing to make this change/PR yourself? Including the migration of existing cache?

**Reply by olira** [reply]
I would first have to fully understand the full implications, but yeah, that could be an option :)
